### ISSUE TITLE
Menu delegates and data-driven menus

### ISSUE THREAD

**Issue by quasado** [electron/electron#528]
Maybe I missunderstood something but trying to overwrite some delegates in a newly created Menu will not call my custom function:

menu.delegate.menuWillShow = function () {
console.log('MY MENU DELEGATE');
}

 menu.delegate.menuWillShow(); // doesn't call my function??

Note that the Menu is created using remote call within a BrowserWindow.

**Comment by hhsnopek** [comment]
@zcbenz Any updates on this? - I know that nw.js has this capability; any chance that we could implement it similar to theirs?

**Comment by dekarguy** [comment]
Has any progress been made on the front of exposing the delegates? I'm trying to build a data driven menu system for an application where the enabled state of menu items are driven by a function and just ran into this issue.

**Comment by JayBizzle** [comment]
Yep, also needing this functionality!

**Comment by petrvecera** [comment]
We would really love to have this functionality in our product!

**Comment by reiberdatschi23** [comment]
Any update on this yet since I've started this topic over a year ago ;) ?

**Comment by chanpory** [comment]
I'd love to see this functionality as well. Perhaps an early holiday present for us?

**Comment by kamesh-a** [comment]
+1

**Comment by jamesmorgan** [comment]
+1 This would be a great feature!

**Comment by mwhearty** [comment]
+1 Could really use dynamic menu functionality in our app.

**Comment by sithika-a** [comment]
+1

**Comment by kirkouimet** [comment]
+1

**Comment by mattui** [comment]
+1

**Comment by bengotow** [comment]
We're really interested in working on this part of the Menu API for N1. The app's menus are growing, and some of the menus we want are used infrequently and are expensive to assemble. (See screenshot.) For our purposes, I think a single change would suffice: in place of an array of items, allow `submenu` to be a function which returns items, which is called when the submenu is about to be displayed.

![image](https://cloud.githubusercontent.com/assets/1037212/14754939/6df12124-0893-11e6-8d63-70fc4991774a.png)

Would be interested to hear other specific use cases for menu delegates that folks have in mind.

**Comment by cicithesquirrel** [comment]
Facing this issue, I found this to work (or seem to):
- `myMenu.clear()` to destroy all menu items,
- `myMenu.append(new MenuItem({...}))` to add new items.

Is it an acceptable by-pass?

What worries me most is that the `clear()` function of [Menu API](https://github.com/electron/electron/blob/master/docs/api/menu.md) is not documented: maybe it is not recommended to call it...

**Comment by lexicalunit** [comment]
This may not be applicable to your use case, @cicithesquirrel, but [in my Atom package](https://github.com/lexicalunit/multi-wrap-guide/blob/master/lib/multi-wrap-guide.coffee#L137) the way I worked around this was to call `dispose()` rather than `clear()` on the menu I wanted to update. Then I add the new menu items just like you're doing. I probably would have used `clear()` though had I known it existed. Would be nice to get `clear()` documented if it's the intended solution for this kind of situation.

**Comment by kennyjwilli** [comment]
+1

**Comment by ncortines** [comment]
+1

**Comment by MarshallOfSound** [comment]
@ncortines @kennyjwilli Please don't comment "+1".  GitHub has reactions now, simply use the "thumbs up" reaction instead of commenting which adds nothing to the conversation and spams everyone's inboxes :+1:

**Comment by nLamprok** [comment]
Not sure if it's the best solution, but I managed to do it the "dirty" way. 

Basically, I sent a message to `ipcMain` from `ipcRenderer` to "rebuild" the menu using the initial template. So whenever I add things to the menus and want to go back to the starting template I simply send a message to `ipcMain`.

It's not ideal, but at least it's something (since .clear() didn't really work):

From renderer:
```js 
ipcRenderer.sendSync('rebuild-menu');
```

At main process:
```js
// cachedTemplate is the initial template used when bootstrapping the application

ipcMain.on('rebuild-menu', (event, arg) => {
  let resetMenu = Menu.buildFromTemplate(cachedTemplate);
  Menu.setApplicationMenu(resetMenu);
  event.returnValue = 'Rebuilt electron menu.';
});
```

NOTE that if you don't provide a `returnValue`,  the application halts.

I know it's not ideal, but since there's no official and documented way of removing things from Menu, I had to basically reset the whole thing on demand.

**Comment by eriedl-kiban** [comment]
I'm trying to implement a "recent items" feature and I'm having difficulties adding/removing MenuItems at run-time. I came across this comment here https://github.com/electron/electron/issues/608#issuecomment-54026327, which led me here. I just wanted to confirm that there is at the moment no way to dynamically add and remove MenuItems or if I'm just doing something wrong. Specifically, I'm trying to add and remove MenuItems from a submenu collection of another MenuItem.

However, `menuItem.submenu.append(new MenuItem{})` seems to add the items, but they only become visible after a state change of the parent, i.e. enabling then disabling it.

Thank you.

**Comment by fantasticsid** [comment]
If you are using `setApplicationMenu` to rebuild and rerender the whole menu every time you want to modify a menu item, it might be worthwhile to look into some memory leak issues. In our application, we use the trick to work around the limitation (can't dynamically change menu label etc). But it seems just allocating a MenuItem (while rebuilding menu) in JS causes memory leaks.

`Menu.clear` is mentioned above as an attempt to clean up, but in practice I found that you need to call `Menu.destroy` as well to further reclaim memory, and I'm not sure it's the proper way to do it / if it claims all memory, as both `clear` and `destroy` are not exposed in API doc.

I can probably look into it a bit more when I get some time.

**Comment by MagicMarvMan** [comment]
+1

**Comment by Anrock** [comment]
Isn't this implemented via https://github.com/electron/electron/pull/11754 ?

**Comment by caseywebdev** [comment]
> Isn't this implemented via #11754 ?

No, the PR you linked implements events for the system menu but does not improve the state of dynamic menu updates, which is what this issue is about.

**Comment by nickensoul** [comment]
> `Menu.clear` is mentioned above as an attempt to clean up, but in practice I found that you need to call `Menu.destroy` as well to further reclaim memory, and I'm not sure it's the proper way to do it / if it claims all memory, as both `clear` and `destroy` are not exposed in API doc.

Thanks, it works, except using `destroy` method leads main process to crash in case of fast multiple updates of menu (doing it in a way where we destroy and generate a new template again).

For now, trying to find a solution beside dummy throttle with value about 500ms or more.

You could reproduce this via creating a function that destroys menu, then generates template and sets to app; then, call it multiple times straight away. Launching app with that code will crash   app. (macOS 10.15, Electron 5.0.3)

**Comment by felixrieseberg** [comment]
Hey everyone! Thank you all so much for giving Electron a try and thinking about ways to make it better. As you can probably tell from how this issue has been open for more than ten years, so far it has never been important enough for someone to close it directly or to otherwise make the resources available for it to happen. 

If you're wondering what you can do to get this into Electron, great news - Electron is truly open source! We, the maintainers, [welcome outside contributions - especially from individuals and companies interested in joining us in maintaining this platform](https://www.electronjs.org/governance).

If you're reading this and are thinking to yourself "Alright, I'll do it!" - good for you! For smaller changes, you can usually open a well-tested and documented PR directly on this repository. For larger changes, I'd recommend that you start with a "[request for comments](https://www.electronjs.org/blog/rfcs)". This will ensure that you don't spend a lot of time on a larger change without getting feedback about your approach first. Alternatively, you could always hire a consultant, either from the maintainer community our outside, to perform the work that's important to you.

In short, I'm closing this issue because we're currently _not planning_ this work, but if you care deeply about this issue, we'd welcome your contribution.
